src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMRanker.scala

```diff
@@ -78,6 +79,20 @@ class LightGBMRanker(override val uid: String)
   }
 
   override def copy(extra: ParamMap): LightGBMRanker = defaultCopy(extra)
+
+  override def prepareDataframe(dataset: Dataset[_], trainingCols: Array[(String, Seq[DataType])],
+                                numWorkers: Int): DataFrame = {
+    val repartitionedDataset = getOptGroupCol match {
+      case None => dataset
+      case Some(groupingCol) => {
+        val df = dataset.repartition(new Column(groupingCol)).cache()
```

Comment:
the changes are technically correct but repartition is very expensive in spark.  Also we are already doing a coalesce here which we should avoid if we are already repartitioning:

https://github.com/Azure/mmlspark/blob/master/src/main/scala/com/microsoft/ml/spark/lightgbm/LightGBMBase.scala#L101

the repartition also has to result in as many partitions as workers for lightgbm to work but that is taken care of by calling coalesce after this.  Maybe for now, we can have a setting on the ranker in order to allow the user to turn off the extra repartitioning done here?  That way, in case the user already has the data partitioned correctly prior to calling lightgbm ranker, execution time won't suffer.  We can leave this as the default behavior though.  In the future it would be nice to figure out how to combine this with the other coalesce/repartition (the latter done in barrier execution mode) in order to avoid doing the expensive operation twice.

Reply:
Maybe we could repartition to the numWorkers? it may be more expensive than just the coalesce done afterwards but I think it would simplify the interface not having an extra flag. Plus, the coalesce will not affect afterwards, since data will be already partitioned to the exact number of partitions desired.

If this solution is not convincing for you I will add the extra setting.

Reply:
Actually I have been thinking, shouldn't Spark DAG planner avoid extra repartitioning if data is already partitioned with the same grouping column policy?